### DISCUSSION TITLE
Variable Mappings in Mixed (aleatory + epistemic) Uncertainty Quantification

### DISCUSSION THREAD

**Discussion by matteo0mamino** [snl-dakota/dakota#52] [Q&A] [answered]
I am trying to run an uncertainty quantification analysis on a flight mechanics problem using Dakota 6.15 on linux. My problem has five uncertain inputs: 
- two are characterized by epistemic uncertainties (descriptors 'WSP' and 'WAZ')
- the remaining three are characterized by aleatory uncertainty ('WAC', 'STA' and 'BL')

The interface and overall coupling with the flight simulator (made up of preprocessing, post processing and driver scripts) are correct and well tested (I have already performed some sensitivity analysis to exclude some parameters from the UQ).

The input file I am experiencing issues with is pasted below:

# ======================================================================================
#  DAKOTA INPUT FILE - dakota_rosenbrock.in
#  This sample Dakota input file optimizes the Rosenbrock function.
#  See p. 95 in Practical Optimization by Gill, Murray, and Wright.

# DAKOTA INPUT FILE FOR FLIGHT FLIGHT0ID (MIXED UNCERTAINTY QUANTIFICATION)

# last reference used dakota_uq_timeseries_sop_optinterf.in

environment
                top_method_pointer = 'EPISTEMIC'
                tabular_data
                               tabular_data_file = 'dakotaoutputs/flightFLIGHT1ID/dakota_psuade.0.dat'         

############################################################
# begin EPISTEMIC specification (wind speed and direction) #
############################################################

method
                id_method = 'EPISTEMIC'
                model_pointer = 'EPIST_M'
                global_interval_est lhs
                               samples = NSAMPLE_EPIST
                               seed = 10

model
                id_model = 'EPIST_M'
                variables_pointer = 'EPIST_V'
                nested
                               sub_method_pointer = 'ALEATORY'

variables
                id_variables = 'EPIST_V'
                continuous_interval_uncertain = 2
                               num_intervals = 1 1 
                               interval_probabilities = 1.0 1.0
                               lower_bounds    0     0     
                               upper_bounds    3     360 
                               descriptors         'WSP' 'WAZ'

#######################################################
# begin ALEATORY specification (mass and CG position) #
#######################################################

method 
                id_method = 'ALEATORY'
                model_pointer = 'ALEAT_M'
                sampling
                               sample_type random
                               samples = NSAMPLE_ALEAT
                               seed = 100

model
                id_model = 'ALEAT_M'
                variables_pointer = 'ALEAT_V'
                responses_pointer = 'ALEAT_R'
                single    
                               interface_pointer = 'ALEAT_I'
                               
variables
                id_variables = 'ALEAT_V'
                active aleatory
                normal_uncertain = 3
                               means                  WACMU  STAMU  BLMU
                               std_deviations WACSTD STASTD BLSTD
                               descriptors         'WAC'  'STA'  'BL'
                continuous_interval_uncertain = 2
                
                
interface
                id_interface = 'ALEAT_I'
                fork
                               analysis_driver = 'flightlab_driver.sh'
                               parameters_file = 'params.in'
                               results_file    = 'results.out'

responses
                id_responses = 'ALEAT_R'
                num_objective_functions = 16
                descriptors = 'XCFCS' 'XBFCS' 'XAFCS' 'XPFCS' 'RIGIDLEFTCOLLECTIVEDEG' 'RIGIDRIGHTCOLLECTIVEDEG' 'ROTTORQUELH' 'ROTTORQUERH' 'ROTPWRLH' 'ROTPWRRH' 'A1LFHUBSIMRX' 'A1RFHUBSIMRX' 'B1LFHUBSIMRX' 'B1RFHUBSIMRX' 'THETA' 'PHI'
                no_gradients
                no_hessians
 
# ======================================================================================

Disclaimer: the response mapping has not been implemented yet in the input file, but that is not the issue.

My idea is to perform an IVP (_https://snl-dakota.github.io/docs/latest_release/users/usingdakota/advanced/advancedmodelrecursions.html_) and obtain an horsetail plot like the one in the link. In particular, I need 'WSP' and 'WAZ' (epistemic uncertain variables) to be specified by the top level method (global_interval_est lhs). Then, I need 'WSP' and 'WAZ' to be passed unmodified to the sub-model (sampling) in order to have 'WAC', 'STA' and 'BL' varying according to the specified MU and STD while keeping 'WSP' and 'WAZ' fixed. Even if this should be a pretty standard case according to the documentation, I am experiencing issues with variable mappings from top to sub-model.

To specify the mappings, I am following the instructions in _https://snl-dakota.github.io/docs/latest_release/users/usingdakota/reference/model-nested-sub_method_pointer.html_. According to the link, I came to the conclusion that option 3 mapping (mapping through variable types) is the best suited for my application.  Thus, I do not explicitly specify any primary or secondary variable mappings, and I leave Dakota to recognize that the continuous_interval_uncertain variables in the sub-model (ALEAT_V) are to be taken from the top level (where two variables of the very same type, 'WSP' and 'WAZ' are specified). Nevertheless, I get the following error running the aforemetioned template:

Input line 68: lower_bounds must be specified for continuous_interval_uncertain.
Input line 69: upper_bounds must be specified for continuous_interval_uncertain.

It is as if Dakota is not recognizing that the continuous_interval_uncertain variables in the sub-model (ALEAT_V) have already been defined.

Can anyone relate to this issue? According to the template and documentation it seems to me that the variable mapping should work fine, but it doesn't. It may be something very simple but I am losing days on this and I was not able to come up with a solution yet. 
Also, if according to the problem I have to solve, you think that another mapping type is more appropriate, please feel free to tell it. 

I remain at disposal for further doubts. 

Thank you for your time, I owe you a beer.

**Comment by jadamstephens** [comment, marked as answer]
Matteo,

You are close to having a working input file. There are two issues that must be corrected.

* Just add bounds to the continuous_interval_variable spec in ALEAT_V. Your intuition is correct that bounds shouldn't be needed, but Dakota isn't that "smart". The input parser reads that variables block and it doesn't know the broader context, that you're doing a mixed UQ study. It only knows that continuous_interval_uncertain variables are required to have bounds.
* You must also have a set of epistemic responses and, in your EPIST_M model, a primary_response_mapping. The epistemic responses are what's being collected by the EPISTEMIC method. Consult the [sub_method_pointer](https://snl-dakota.github.io/docs/6.18.0/users/usingdakota/reference/model-nested-sub_method_pointer.html) keyword documentation to learn about the primary response mapping.

**Reply by matteo0mamino** [reply]
Thank you very much Adam! This solved all the issues.
